File: Problems/WordLadderII.py
import collections

class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        words=set(wordList)
        if endWord not in words:return[]
        words.add(beginWord)
        
        aux=defaultdict(set)
        
        for word in words:
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    neww=word[:i]+c+word[i+1:]
                    if neww!=word and neww in words:
                        aux[word].add(neww)
                        aux[neww].add(word)
        
        
        distance=defaultdict(int)
        vis=set([beginWord])
        queue=deque([(beginWord,1)])
        while queue:
            curr,d=queue.popleft()
            distance[curr]=d
            for i in aux[curr]:
                if i not in vis:
                    queue.append((i,d+1))
                    vis.add(i)
        
        result=[]
        stack=[(beginWord,[beginWord])]
        while stack:
            curr,path=stack.pop()
            if curr==endWord and len(path)==distance[endWord]:
                result.append(path)
            for i in aux[curr]:
                if distance[i]>distance[curr]:
                    stack.append((i,path+[i]))
        
        return result

File: Problems/wordLadderII.go
func differencesChar(x, y string ) int {
    result := 0
    for iter, _ := range x{
        if x[iter] != y[iter]{
            result++;
        }
    }
    return result
}

func findLadders(beginWord string, endWord string, wordList []string) [][]string {
    hashMap := map[string]int{beginWord:0} 
    q := make([]string, 0) 
    q = append(q, beginWord)
    for len(q) > 0 { 
        currentWord := q[0]
        q = q[1:] 
        for _, nextWord := range wordList { 
            if _, have := hashMap[nextWord]; !have && differencesChar(currentWord, nextWord) == 1 { 
                q = append(q, nextWord) 
                hashMap[nextWord] = hashMap[currentWord] + 1
            }
        }
    }
    
    result := [][]string{}
    if _, have := hashMap[endWord]; !have {
        return result
    }

    var dfs func(currentWord string, current []string) 
    dfs = func(currentWord string, current []string) {
        current = append(current, currentWord)
        if (hashMap[currentWord] == 1) { 
            current = append(current, beginWord) 
            aux := current
            for i, j := 0, len(aux)-1; i < j; i, j = i + 1, j - 1 {
                aux[i], aux[j] = aux[j], aux[i]
            }
            result = append(result, aux)
            current = current[:len(current)-2]
            return;
        }           
        for _, nextWord := range wordList {
            if _, have := hashMap[nextWord]; have && differencesChar(currentWord, nextWord) == 1 && hashMap[currentWord] == hashMap[nextWord] + 1 {
                dfs(nextWord, current)
            }
        }        
        current = current[:len(current)-1]
    }
    dfs(endWord, []string{})
    
    return result
}
